core/math/dynamic_bvh.cpp

```diff
@@ -0,0 +1,400 @@
+			node->childs[1]->parent = node;
+			return (node);
+		} else {
+			_bottom_up(leaves, p_count);
+			return (leaves[0]);
+		}
+	}
+	return (leaves[0]);
+}
+
+DynamicBVH::Node *DynamicBVH::_node_sort(Node *n, Node *&r) {
+	Node *p = n->parent;
+	ERR_FAIL_COND_V(!n->is_internal(), nullptr);
+	if (p > n) {
+		const int i = n->get_index_in_parent();
+		const int j = 1 - i;
+		Node *s = p->childs[j];
+		Node *q = p->parent;
+		ERR_FAIL_COND_V(n != p->childs[i], nullptr);
+		if (q)
+			q->childs[p->get_index_in_parent()] = n;
+		else
+			r = n;
+		s->parent = n;
+		p->parent = n;
+		n->parent = q;
+		p->childs[0] = n->childs[0];
+		p->childs[1] = n->childs[1];
+		n->childs[0]->parent = p;
+		n->childs[1]->parent = p;
+		n->childs[i] = p;
+		n->childs[j] = s;
+		SWAP(p->volume, n->volume);
+		return (p);
+	}
+	return (n);
+}
+
+void DynamicBVH::clear() {
+	if (bvh_root) {
+		_recurse_delete_node(bvh_root);
+	}
+	lkhd = -1;
+	opath = 0;
+}
+
+void DynamicBVH::optimize_bottom_up() {
+	if (bvh_root) {
+		LocalVector<Node *> leaves;
+		_fetch_leaves(bvh_root, leaves);
+		_bottom_up(&leaves[0], leaves.size());
+		bvh_root = leaves[0];
+	}
+}
+
+void DynamicBVH::optimize_top_down(int bu_treshold) {
+	if (bvh_root) {
+		LocalVector<Node *> leaves;
+		_fetch_leaves(bvh_root, leaves);
+		bvh_root = _top_down(&leaves[0], leaves.size(), bu_treshold);
+	}
+}
+
+void DynamicBVH::optimize_incremental(int passes) {
+	if (passes < 0)
+		passes = total_leaves;
+	if (bvh_root && (passes > 0)) {
+		do {
+			Node *node = bvh_root;
+			unsigned bit = 0;
+			while (node->is_internal()) {
+				node = _node_sort(node, bvh_root)->childs[(opath >> bit) & 1];
+				bit = (bit + 1) & (sizeof(unsigned) * 8 - 1);
+			}
+			_update(node);
+			++opath;
+		} while (--passes);
+	}
+}
+
+DynamicBVH::ID DynamicBVH::insert(const AABB &p_box, void *p_userdata) {
+	Volume volume;
+	volume.min = p_box.position;
+	volume.max = p_box.position + p_box.size;
+
+	Node *leaf = _create_node_with_volume(nullptr, volume, p_userdata);
+	_insert_leaf(bvh_root, leaf);
+	++total_leaves;
+
+	ID id;
+	id.node = leaf;
+	return id;
+}
+
+void DynamicBVH::_update(Node *leaf, int lookahead) {
+	Node *root = _remove_leaf(leaf);
+	if (root) {
+		if (lookahead >= 0) {
+			for (int i = 0; (i < lookahead) && root->parent; ++i) {
+				root = root->parent;
+			}
+		} else
+			root = bvh_root;
+	}
+	_insert_leaf(root, leaf);
+}
+
+void DynamicBVH::update(const ID &p_id, const AABB &p_box) {
+	ERR_FAIL_COND(!p_id.is_valid());
+	Node *leaf = p_id.node;
+	Node *base = _remove_leaf(leaf);
+	Volume volume;
+	volume.min = p_box.position;
+	volume.max = p_box.position + p_box.size;
+	if (base) {
+		if (lkhd >= 0) {
+			for (int i = 0; (i < lkhd) && base->parent; ++i) {
+				base = base->parent;
+			}
+		} else
+			base = bvh_root;
+	}
+	leaf->volume = volume;
+	_insert_leaf(base, leaf);
+}
+
+void DynamicBVH::remove(const ID &p_id) {
+	ERR_FAIL_COND(!p_id.is_valid());
+	Node *leaf = p_id.node;
```

Comment:
What happens if the provided ID is not a leaf node? Should we guard against it?

Reply:
should be ok since the pointer is encapsulated in ID

Reply:
The pointer inside ID could thechnically reference an internal node, but I guess the rest of the API doesn't really allow it.

In any case guarding against this would only be useful for debugging purposes.